File: web/src/lib/parse.ts
export type ParsedAccount = { email: string; password: string };

const emailPass = /^([^\s:]+@[^\s:]+)(?::|----)(.+)$/i;

export function parseAccounts(text: string): ParsedAccount[] {
  const seen = new Set<string>();
  const out: ParsedAccount[] = [];
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#")) continue;
    let email = "";
    let password = "";
    if (line.includes("----")) {
      const idx = line.indexOf("----");
      email = line.slice(0, idx).trim();
      password = line.slice(idx + 4).trim();
    } else {
      const match = line.match(emailPass);
      if (!match) continue;
      email = match[1];
      password = match[2].trim();
    }
    if (!looksLikeEmail(email) || !password) continue;
    const key = email.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    out.push({ email, password });
  }
  return out;
}

function looksLikeEmail(value: string) {
  const at = value.lastIndexOf("@");
  return at > 0 && value.slice(at + 1).includes(".");
}

File: server/internal/importparse/parse.go
package importparse

import (
	"regexp"
	"strings"
)

type Account struct {
	Email    string `json:"email"`
	Password string `json:"password"`
	Raw      string `json:"raw,omitempty"`
}

var emailPass = regexp.MustCompile(`(?i)^([^\s:]+@[^\s:]+)(?::|----)(.+)$`)

func Parse(text string) []Account {
	lines := strings.Split(strings.ReplaceAll(text, "\r\n", "\n"), "\n")
	out := make([]Account, 0, len(lines))
	seen := map[string]struct{}{}
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		email, password, ok := splitLine(line)
		if !ok {
			continue
		}
		key := strings.ToLower(email)
		if _, exists := seen[key]; exists {
			continue
		}
		seen[key] = struct{}{}
		out = append(out, Account{Email: email, Password: password, Raw: line})
	}
	return out
}

func splitLine(line string) (string, string, bool) {
	if strings.Contains(line, "----") {
		email, password, ok := strings.Cut(line, "----")
		email = strings.TrimSpace(email)
		password = strings.TrimSpace(password)
		if !ok || !looksLikeEmail(email) || password == "" {
			return "", "", false
		}
		return email, password, true
	}
	match := emailPass.FindStringSubmatch(line)
	if len(match) != 3 {
		return "", "", false
	}
	return strings.TrimSpace(match[1]), strings.TrimSpace(match[2]), true
}

func looksLikeEmail(value string) bool {
	at := strings.LastIndex(value, "@")
	if at <= 0 || at == len(value)-1 {
		return false
	}
	return strings.Contains(value[at+1:], ".")
}
